### PR TITLE
Improve Random Generation of Types and Values

### PR DESCRIPTION
**Pull request by danking**
Addresses #1894

There's two big things I want critique on (besides, of course, the implementation): what the samples look like now and how I got there.

# What Samples Look Like Now

I recommend drawing types and values of a type as follows. Different users might more or less space used on types.

```scala
   val g = for {
      s <- Gen.size
      // prefer smaller type and bigger values
      fraction <- Gen.choose(0.1, 0.3)
      x = (fraction * s).toInt
      y = s - x
      _ = println(s"sizes: $x, $y")
      t <- Type.genStruct.resize(x)
      v <- t.genNonmissingValue.resize(y)
    } yield (t, v)
```

The outputs of this generator for `size` = 1000 look like the following. The generator does a much better job of exploring the type's value space. However, I need to counteract the tendency of the type generator to create nested sequences.

## Example Draw 1
```scala
t Struct {
    aj0iFnxi: Array[Set[Array[Dict[Empty, Dict[Array[Call], Call]]]]],
    dA0IS78I: Set[Empty],
    VeGLA3v: Array[Array[Struct {
        C: Empty,
        sHjMXnj: Boolean,
        G: Call
    }]],
    Ni: Struct {
        HCrSI: Empty,
        nJt7: Boolean
    },
    nxb8CkLI: Int,
    y4mYv_DvH: Dict[Array[Genotype], Variant]
}
v [ WrappedArray( Set( WrappedArray(null)
                     , WrappedArray()
                     , WrappedArray(Map(null -> Map(WrappedArray(2) -> 2)), Map(null -> Map(WrappedArray() -> 1)), Map(null -> Map(WrappedArray(1) -> 2)))
                     )
                , Set(WrappedArray(Map(null -> null), Map(null -> Map(WrappedArray(2) -> 31, WrappedArray(0) -> 0))))
                , Set(WrappedArray(Map(null -> Map()), Map(null -> null)))
                , Set( WrappedArray(Map())
                     , WrappedArray(Map(null -> Map()))
                     , WrappedArray(Map(null -> null))
                     , WrappedArray()
                     , WrappedArray(null, Map(null -> Map(WrappedArray() -> null))))
                , Set()
                , Set( WrappedArray(Map(null -> Map(WrappedArray() -> null)))
                     , WrappedArray()
                     , WrappedArray(Map(null -> Map()))
                     , WrappedArray(Map(null -> Map()), Map(null -> Map(WrappedArray() -> 2))))
                , Set(WrappedArray())
                , Set( WrappedArray()
                     , WrappedArray(Map(null -> Map())))
                , Set()
                , Set( WrappedArray(Map(null -> null))
                     , WrappedArray(Map(null -> Map())))
                , Set( WrappedArray(Map(null -> Map(WrappedArray(null) -> 1)), Map(null -> Map(WrappedArray(5) -> 2)))
                     , WrappedArray(Map()))
                , Set()
                , Set()
                , Set(WrappedArray()
                     , WrappedArray(Map(null -> Map())))
                , Set()
                , Set( null
                     , WrappedArray())
                , Set( WrappedArray(Map(null -> Map(WrappedArray(1) -> 19)))
                     , WrappedArray(Map(null -> Map(WrappedArray(0) -> 2)), Map(null -> Map()))
                     , WrappedArray(Map(null -> Map()))
                     , WrappedArray(Map(null -> Map(WrappedArray(2) -> null)), Map(null -> Map(WrappedArray() -> 17)), Map(null -> Map(WrappedArray() -> 1))))
                , Set(WrappedArray(Map(null -> Map(WrappedArray(null) -> 1)), Map(null -> Map(WrappedArray(2) -> 0, WrappedArray() -> null)), Map(), Map()))
                , Set(WrappedArray(Map(null -> Map()), Map(null -> Map()), Map(null -> Map(WrappedArray(0) -> 0))))
                , Set( WrappedArray()
                     , WrappedArray(Map(null -> Map(WrappedArray(0) -> null))))
                , Set(WrappedArray(Map()))
                , Set( WrappedArray(Map())
                     , WrappedArray(Map(null -> Map(WrappedArray() -> 0)))
                     , WrappedArray(Map(null -> Map())))
                , Set()
                , Set( WrappedArray()
                     , WrappedArray(Map(null -> Map(WrappedArray(null) -> 2)))
                     , WrappedArray(Map(null -> Map())))
                , Set( WrappedArray(Map(null -> Map(WrappedArray() -> 0)))
                     , WrappedArray(Map(), Map(null -> Map()), Map(), Map(null -> Map(WrappedArray(0) -> null))))
                )
  , Set(null)
  , WrappedArray( WrappedArray(null, [null,false,1], [null,true,6])
                , WrappedArray()
                , WrappedArray([null,false,0], [null,true,1])
                , WrappedArray()
                , null
                , WrappedArray()
                , null
                , WrappedArray(null, [null,false,0], null)
                , WrappedArray(null, null, null, [null,false,null])
                , WrappedArray(null, [null,true,null])
                , WrappedArray([null,true,null])
                , WrappedArray([null,true,1])
                , WrappedArray([null,false,null], null, null)
                , WrappedArray([null,null,null]))
  , null
  , 2147483647
  , Map( WrappedArray(1/1:.:.:3:PL=4,328,0) -> Xtfuo:983235984:G:ACT
       , WrappedArray() -> null
       , WrappedArray( 0/0:.:.:.:GP=0.835968017578125,0.00244140625,0.161590576171875
                     , 0/0:401640437,122557639:1151958249:2098:PL=0,29,43
                     , 1/1:30,8:65:128:PL=419,997,0
                     ) -> zEeTaoQ:931965768:AGA:*
       , WrappedArray(1/1:.:357602351:3000:PL=53,81,0) -> bu8qWa3:440101875:C:CAACG,G,TAAAT,GT,AC,GC,A,TTGCGA
       )
  ]
```
## Example Draw 2
```scala
t Struct{
    W: Empty,
    VH6KA0: Array[Empty],
    Y6b: Array[Dict[Interval, Set[Interval]]],
    gvU1NyKqvR: Set[Dict[Set[Double], Locus]],
    wY: Struct{
        nT: Variant,
        L_9BUfa0z8D: Empty
    },
    eiN: Array[Variant],
    bj36QLPq: Struct{
        EPMpm: Long,
        ohcAn: Interval
    },
    y26WlwsMcu: Dict[Locus, Double],
    LpE: Array[String],
    UroYb: String,
    dk: Struct{
        VONVG8X9R: Boolean,
        eo6fz1YMhtD: Empty
    },
    wi: Interval,
    JK0osYJHRz: Array[Dict[Long, Long]]
}
v [ null
  , WrappedArray(null, null, null)
  , WrappedArray( Map(21:474864761-21:2142353440 -> Set(12:291332689-21:2122081729))
                , Map(6:64658535-18:852482744 -> Set(4:920440399-20:99044206), 1:1996448901-Y:2076322341 -> Set(21:1147405636-Y:1021924735))
                , Map()
                , Map(14:1359868957-19:1770937039 -> Set(8:2031139633-X:1928544108, 4:384201545-12:1909645017, 10:1688101834-20:1908406452))
                , Map()
                , Map()
                , Map()
                , Map(7:2039691980-20:185934176 -> Set(3:1071992365-14:349831984), 9:478984210-12:1771267836 -> Set(4:650211214-11:1348097899), 5:520593968-5:1367289651 -> Set(null))
                , Map(1:2111004582-18:615377082 -> Set(18:230351014-19:1831314528))
                , Map(null -> Set(3:1954841637-12:306284522), 1:92294299-7:1257997486 -> Set(7:1713008336-13:1996067271), 2:880982333-10:1907495418 -> Set())
                , Map()
                , Map(6:560306885-18:648427852 -> Set())
                , Map(2:1018997744-15:1183423256 -> Set(4:227078176-X:662176672, 5:1587706484-15:800371306))
                , Map(10:1449101177-18:1831584143 -> Set(18:184015519-19:925642304))
                , Map(18:140216687-18:1914060841 -> Set())
                , Map()
                , Map(7:542188076-22:1646282667 -> null)
                , Map(5:1899824081-15:1106733079 -> Set(19:622233834-X:1479799008), 6:2011592907-16:1486331302 -> Set(), 19:206220313-MT:218599517 -> Set(3:1241210118-14:172955931), 22:1097142563-MT:1684780520 -> Set(), 16:1094346819-21:374152140 -> Set(17:298725861-Y:276538111))
                , Map(2:1576674560-19:2051426361 -> Set(3:1729279166-4:1468182755))
                )
  , Set( Map(Set(37.286923408013536) -> 14:236896985, Set(-15.021445198381315) -> 3:2086181046, Set(18.144202437411707) -> 11:43701721, Set(4.9E-324) -> 15:152457683)
       , Map(null -> 1:930696534)
       , Map(Set(4.9E-324) -> 19:1442338618)
       , Map(Set(-36.763366959632954) -> 22:672685435, Set(-1.7976931348623157E308) -> 9:1705107082)
       , Map(Set() -> 1:186365173, Set(-6.398293809835415E307) -> 11:26598828)
       , Map()
       , Map(Set() -> 19:759113415, Set(0.0, 1.9948878010656322E307, 22.324353631389556) -> 7:1349934030)
       , Map(Set() -> 6:10152252, Set(1.7976931348623157E308) -> 8:1499191709, Set(4.607043291623889E307) -> 5:316220126, Set(1.0) -> 20:1314749173, Set(4.9E-324) -> 10:1106413461)
       )
  , [D3YSSo:1677841928:TA:*,null]
  , WrappedArray( AuXYPeE9g:1581545128:G:GG
                , VPnCju:2025837886:CAG:TGG
                , GZoiw:11124631:TG:A
                , C5c6HNsTe:1539490834:TA:AAC,C,AGC,GAA,GTCA,ACGC,CCT,GG
                , z7jhYG8STJ:523877649:A:T,GGACA,G,*,TC,AAAC,AA,GG,AGA
                , y0eEFwipq98:835986978:GT:TGTGT
                , E4al5h:482756775:C:CGAC
                , HAJ1jrSuaoq:1724809203:GA:TC
                , w:1098499131:GT:GGT
                , dHCyUjxqhr:1258701198:TA:AGC,CC,GTT,TGA,ACATGGA,GC,A,TGAGA,TAAC
                )
  , [-16,14:878485900-MT:1655450544]
  , Map(4:1001834270 -> -1.4590773156850653E308, 10:2024397357 -> 41.20125040141795)
  , WrappedArray(-E, x', , mZW}Xaz, '(<vn:R, wNtR{u'T, h, , 0)C, PJ), :Q-#J7)N39j, g, , null)
  , k
  , [true,null]
  , 11:422731466-12:1318927466
  , WrappedArray( Map(-4362806367897070770 -> -1)
                , Map(-64 -> 37, 8932649469340641424 -> -9223372036854775808, -2581551729758561812 -> 4866402986430542844, -21 -> -53)
                , null
                , Map(48 -> -1, -9203674311477531463 -> -20))
  ]
```

# Example Draw 3
```scala
t Struct{
    YiG6N: Set[Set[Struct{
        hAu0wTu: Empty
    }]],
    tGFebQm0H: Array[Array[Array[Empty]]],
    _t24kfMLN: Long,
    eSPJnjhs3: Array[Dict[Boolean, Dict[Interval, Empty]]],
    B: Float,
    pXm: Set[Genotype],
    JRsi9GV: Array[Struct{
        k0sLX9N: Empty
    }],
    pMq: Set[Dict[Set[Call], Empty]],
    _: Empty,
    J: AltAllele,
    zo3Fqqd: Variant,
    gxw1J4E2s: Double,
    NB: Struct{
        aD5P8: Dict[String, Variant]
    }
}

v [ Set(Set([null]), Set())
  , WrappedArray( WrappedArray(WrappedArray(null, null, null, null, null, null))
                , WrappedArray(WrappedArray(null, null, null), WrappedArray()))
  , 7465031920428684276
  , WrappedArray( Map(false -> Map())
                , Map(false -> Map(X:1505701634-MT:413708900 -> null, 2:908273550-4:1601452326 -> null))
                , Map()
                , Map(true -> Map(), false -> Map())
                , Map()
                , Map(true -> Map())
                , Map(false -> Map(2:1696581492-16:1148530363 -> null))
                , Map()
                , null
                , Map()
                , Map(false -> Map(11:1204664327-Y:452748442 -> null), true -> Map(1:759029796-8:660188853 -> null))
                , Map()
                , Map(true -> Map(6:1436710378-14:172736518 -> null)))
  , 0.0
  , Set( ./.:.:.:.:GP=.
       , ./.:666232109,323735949:1572839794:5865:PL=414367811,46987628,0
       , ./.:37,16:69:123:PL=0,482,280
       , 0/0:.:.:.:GP=1.0,0.0,0.0
       , ./.:42,44:91:99:PL=453,0,746
       , 0/1:.:.:.:GP=0.448516845703125,0.551483154296875,0.0
       , 0/0:22,45:80:90:PL=0,535,934
       )
  , null
  , Set( null
       , Map(Set() -> null, Set(null) -> null)
       , Map(Set(2) -> null, Set() -> null)
       , Map()
       , Map(Set(0) -> null, Set() -> null, Set(0, 2) -> null)
       , Map(Set(1) -> null)
       , Map(Set(0) -> null))
  , null
  , AAT/GATGA
  , bm7TK_:989763744:CAGT:GG
  , 4.9E-324
  , [Map(null -> HaUqd5y:234050762:A:G
        , x7d=h,D -> z:141633474:CTC:AGAT,AGT,AGCAA,AACG,TA,TCTG,GT,GAT,GC
        , rqS3:WB'I -> null
        , a?7o5Z% -> Nmm:1780043833:AG:CAG
        , HRt -> hrG6:997661503:CT:TC
        ,  "t:X"d_. -> Z1JGKeEhh:27225644:TTGTAG:GGG
        , ZwU[EK| -> wLtpOSe:892463952:CG:AAC,C,AG,G,CC,ATC,A
        , ;jOwvw= -> EcrHdUW9Bpc:1201822623:TT:G)]]
```

# How it Samples Now

 1. Empty values of any type are now generated with a probability of 1 in 20 rather than 1 in 2.

 2. Whenever we partition the `size` of a random generator, we partition using a [Dirichlet-Multinomial distribution](https://en.wikipedia.org/wiki/Dirichlet-multinomial_distribution) with all α_n set to `parts`, which tends towards a uniformly distributed vector. Upon writing this PR, I realize it might be more desirable to set all α_n to 1. Commentary welcome.

 3. When we choose the length of a random sequence, we choose it using the [Beta-Binomial Distribution](https://en.wikipedia.org/wiki/Beta-binomial_distribution) with `n` set to the maximum sequence size (usually the Gen's `size`), α set to `3` and β set to `6 * ln(n + 0.01)`. My choice of β is motivated by keeping the mass away from `0` when `n` is small. In particular, I'd like the mean to be greater than or equal to one, so `nα/(α+β) >= 1` so for, say, n=4, `β/α <= 3`. For large values of `n` I want values closer to zero than given by α=3, β=4.

I really want commentary on how to choose sequence lengths. Choosing from a Beta-Binomial with a beta weighted by the natural log (shifted by 0.01 to avoid ln(1) = 0) seems very unnatural.

Here's α=3, β=6, at n=4 and n=100:

screen shot 2017-06-05 at 2 40 04 pm
screen shot 2017-06-05 at 2 40 18 pm

Note that the mean for n=100 is close to 35, pretty big.

Now here's α=3, β=6*ln(n+0.01), at n=4 and n=100:

screen shot 2017-06-05 at 2 40 54 pm
screen shot 2017-06-05 at 2 40 34 pm

The n=4 case doesn't change much but the n=100 case biases towards smaller values, which is generally what we want for random sequences (if any algorithm works on empty, size 1 and size 10, it probably works at larger sizes, modulo memory issues).

### CODE REVIEW

**Comment by danking** [comment]
A few new test failures coming from better generators, I assume. IBD one looks like a weird corner case where we differ from plink when there are only three variants. I'll look at the rest soon.

**Comment by tpoterba** [comment]
this is awesome. I'll look at it as soon as I can get a chance, but the fact that you're catching undiscovered bugs means you did something right!

File: src/main/scala/is/hail/sparkextras/OrderedRDD.scala [hunk 1]
```diff
@@ -148,26 +142,28 @@ object OrderedRDD {
...
       }
 
-      while (it.hasNext && overlaps(sortedKeyInfo(it.head)))
+      while (it.hasNext && overlaps(sortedKeyInfo(it.head))) {
+        anyOverlaps = true
         indicesBuilder += it.next()
+      }
 
-      if (it.hasNext && sortedKeyInfo(it.head).min == max)
+      if (it.hasNext && sortedKeyInfo(it.head).min == max) {
```
Comment (cseed) [inline src/main/scala/is/hail/sparkextras/OrderedRDD.scala]: I think you can nuke `overlaps` and the previous loop and make this loop `while (it.hasNext && ....min == max)`.

File: src/main/scala/is/hail/sparkextras/OrderedRDD.scala [hunk 2]
```diff
@@ -123,15 +115,17 @@ object OrderedRDD {
...
-  def rangesAndAdjustments[PK, K, V](sortedKeyInfo: Array[PartitionKeyInfo[PK]])
-    (implicit kOk: OrderedKey[PK, K], vct: ClassTag[V]): (IndexedSeq[Array[Adjustment[(K, V)]]], Array[PK]) = {
+  def rangesAndAdjustments[PK, K, V](sortedKeyInfo: Array[PartitionKeyInfo[PK]], sortedness: Int)
+    (implicit kOk: OrderedKey[PK, K], vct: ClassTag[V]): (IndexedSeq[Array[Adjustment[(K, V)]]], Array[PK], Int) = {
     import kOk._
     import scala.Ordering.Implicits._
 
     val rangeBounds = mutable.ArrayBuilder.make[PK]
     val adjustmentsBuffer = new mutable.ArrayBuffer[Array[Adjustment[(K, V)]]]
     val indicesBuilder = mutable.ArrayBuilder.make[Int]
```
Comment (danking) [inline src/main/scala/is/hail/sparkextras/OrderedRDD.scala]: I made this change.

File: src/main/scala/is/hail/sparkextras/OrderedRDD.scala (hunk 1 above)
Comment (danking) [inline src/main/scala/is/hail/sparkextras/OrderedRDD.scala]: @tpoterba edited this so I'll let him respond and/or make the change.

**Comment by danking** [comment]
@cseed back to you.

**Comment by danking** [comment]
@cseed back to you test hopefully pass now

File: src/main/scala/is/hail/sparkextras/OrderedRDD.scala (hunk 1 above)
Comment (cseed) [inline src/main/scala/is/hail/sparkextras/OrderedRDD.scala]: bump

Comment (tpoterba) [inline src/main/scala/is/hail/sparkextras/OrderedRDD.scala]: doing it now

Comment (tpoterba) [inline src/main/scala/is/hail/sparkextras/OrderedRDD.scala]: here's my replacement:
```scala
      var continue = true
      while (continue && it.hasNext && sortedKeyInfo(it.head).min == max) {
        anyOverlaps = true
        if (sortedKeyInfo(it.head).max == max)
          indicesBuilder += it.next()
        else {
          indicesBuilder += it.head
          continue = false
        }
      }
```

**Comment by tpoterba** [comment]
I did my part - should be good to go?

**Comment by danking** [comment]
@cseed back to you.
